### PR TITLE
[WIP][RFC] CBMs and allies: allow allies to install on you or each other

### PR DESCRIPTION
**Pull request by mlangsdorf**
Very much a work in process, but pushing it up for more review. I might have hacked some things in here that I shouldn't have.

It works, mostly, afaict from quick testing. At least, if you position an NPC on the autodoc couch and install a CBM, you don't have the CBM and if you try to uninstall it, it shows up on the NPC.

For some reason, the code to consume the CBM after the install is failing, even though the
```
                 std::vector<item_comp> comps;
                 comps.push_back( item_comp( it->typeId(), 1 ) );
                 p.consume_items( comps );
```
sequence is something I didn't change.

Open question for which I could use some suggestions: how to display installed CBMs on an NPC, both in the @ menu and in the debug menu?

Todo:
1. Add NPC CBM display
2. Add debug functions to add/remove CBMs from NPCs
3. Whitelist CBMs for NPCs
4. Spawn NPCs with CBMs.

Some of those will be in separate PRs.

### CODE REVIEW

**Comment by mlangsdorf** [comment]
Just realized: the CBMs are being consumed properly, but the anesthesia kits aren't, because patient.introduce_into_anesthesia() consumes the kits but the patient isn't the player and may not have them. Should be easy to fix.

Still need to figure out how to display CBMs on NPCs.

File: src/bionics.cpp [hunk 1]
```diff
@@ -887,42 +888,29 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
     }
 
     // removal of bionics adds +2 difficulty over installation
-    int chance_of_success;
-    if( skill_level != -1 ) { // NPC surgeon removal
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Why did you remove this case? NPC surgeon isn't using autodoc to remove bionics, so he doesn't need to configure it and therefore doesn't need to check for computers skill.

File: src/bionics.cpp [hunk 2]
```diff
@@ -887,42 +888,29 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
...
-        ///\EFFECT_INT increases chance of success removing bionics with unspecified skill level
-        chance_of_success = bionic_manip_cos( int_cur,
-                                              get_skill_level( skilll_firstaid ),
-                                              get_skill_level( skilll_computer ),
-                                              get_skill_level( skilll_electronics ),
-                                              difficulty + 2, autodoc );
-    }
+    int chance_of_success = installer.bionic_manip_cos( skilll_firstaid,
+		                                        skilll_computer,
+							skilll_electronics,
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Weird indentation here.

File: src/bionics.cpp [hunk 3]
```diff
@@ -958,28 +948,10 @@ bool player::install_bionics( const itype &type, int skill_level, bool autodoc )
...
-                                              difficult, autodoc );
-    } else { // Manual self-installation
-        ///\EFFECT_INT increases chance of success installing bionics with unspecified skill level
-        chance_of_success = bionic_manip_cos( int_cur,
-                                              get_skill_level( skilll_electronics ),
-                                              get_skill_level( skilll_firstaid ),
-                                              get_skill_level( skilll_mechanics ),
-                                              difficult, autodoc );
-    }
+    int chance_of_success = installer.bionic_manip_cos( skilll_firstaid,
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: The same here. NPC surgeons are using unspecified skill set for installing, manual installing uses different skill set, not the same as while using autodoc.

File: src/iexamine.cpp [hunk 4]
```diff
@@ -3530,29 +3574,28 @@ void iexamine::autodoc( player &p, const tripoint &examp )
...
+                _( "You prepare to install the %1$s on %2$s." ),
+                _( "<npcname> prepares to install the %1$s on %2$s." ),
+                it->display_name().c_str(), patient.name );
+            if( patient.has_bionic( bid ) ) {
+                popup( _( "%1$s already installed this bionic."  ), pronoun );
                 return;
-            } else if( bid->upgraded_bionic && !p.has_bionic( bid->upgraded_bionic ) ) {
-                popup( _( "You have no base version of this bionic to upgrade." ) );
+            } else if( bid->upgraded_bionic && !patient.has_bionic( bid->upgraded_bionic ) ) {
+                popup( _( "%1$s no base version of this bionic to upgrade." ), pronoun );
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Please add translation comments here about what this `%1$s` here and below supposed to mean.

File: src/iexamine.cpp [hunk 5]
```diff
@@ -3530,29 +3574,28 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             const itype *itemtype = it->type;
             const bionic_id &bid = itemtype->bionic->id;
 
-            if( p.has_bionic( bid ) ) {
-                popup( _( "You have already installed this bionic."  ) );
+	    installer.add_msg_player_or_npc(
+                _( "You prepare to install the %1$s on %2$s." ),
+                _( "<npcname> prepares to install the %1$s on %2$s." ),
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: This line implies that NPCs are able to install bionics on other patients. But as I understand it, this PR allows only player characters to install CBMs on NPCs.

File: src/bionics.cpp [hunk 6]
```diff
@@ -811,21 +809,24 @@ int bionic_manip_cos( int p_int, int most_important_skill, int important_skill,
...
+		                          _( "<npcname> prepares for surgery." ) );
     }
 
     // People trained in using the Autodoc gain an additional advantage towards using it
-    if( g->u.has_trait( trait_PROF_AUTODOC ) ) {
+    if( has_trait( trait_PROF_AUTODOC ) ) {
         pl_skill += 7;
-        add_msg( m_neutral, _( "A lifetime of Autodoc use has taught you a thing or two..." ) );
+        add_msg_player_or_npc( m_neutral,
+		               _( "A lifetime of Autodoc use has taught you a thing or two..." ),
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Weird indentation here.

File: src/bionics.cpp [hunk 7]
```diff
@@ -933,7 +921,9 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
                           pgettext( "memorial_female", "Removed bionic: %s." ),
                           bionics[b_id].name.c_str() );
         // until bionics can be flagged as non-removable
-        add_msg( m_neutral, _( "You jiggle your parts back into their familiar places." ) );
+        add_msg_player_or_npc( m_neutral,
+		               _( "You jiggle the parts back into their familiar places." ),
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Weird indentation here.

File: src/bionics.cpp [hunk 8]
```diff
@@ -1043,116 +1015,77 @@ bool player::install_bionics( const itype &type, int skill_level, bool autodoc )
...
-            break;
-    }
+    add_msg( m_neutral, _( "The installation is a failure." ) );
 
-    if( u->has_trait( trait_PROF_MED ) || u->has_trait( trait_PROF_AUTODOC ) ) {
+    if( installer.has_trait( trait_PROF_MED ) || installer.has_trait( trait_PROF_AUTODOC ) ) {
         //~"Complications" is USian medical-speak for "unintended damage from a medical procedure".
-        add_msg( m_neutral, _( "Your training helps you minimize the complications." ) );
+        add_msg_player_or_npc( m_neutral,
+                            _( "Your training helps you minimize the complications." ),
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Weird indentation here.

File: src/iexamine.cpp [hunk 9]
```diff
@@ -3474,6 +3477,47 @@ void iexamine::climb_down( player &p, const tripoint &examp )
...
+            adjacent_couch = true;
+            if( p.pos() == couch_loc ) {
+                return p;
+            }
+            for( const npc *e : g->allies() ) {
+                if( e->pos() == couch_loc ) {
+                   return  *g->critter_by_id<player>( e->getID() );
+                }
+            }
+	}
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Incorrect indentation here.

**Comment by AMurkin** [comment]
Indentation problems are tabs mostly.

File: src/bionics.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/bionics.cpp]: I was hoping to get rid of the special case: there's always an installer who may or may not be the PC. If some installers should have extra skill in some instances, I'd rather flag it in the skill calculation.

Doing it my way also removes an extra copy of the code to calculate skill in the failure path.

**Comment by mlangsdorf** [comment]
tab issues are some misconfigured vi thing on my side. I'll fix it up before final submission.

**Comment by mlangsdorf** [comment]
NPC Tonza successfully installs a Joint Torsion CBM on himself and goes into sedation.
allycbminstallonally1

NPC Tonza unsuccessfully installs a Cerebral Booster CBM on himself
allycbminstallonally2
I still need to sort out some messages on the error path, but things are getting better.

**Comment by mlangsdorf** [comment]
lua/catabidings.cpp is in the .gitignore, but I need to edit it because I changed the install_bionics call. What's the correct way to handle that?

File: src/bionics.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/bionics.cpp]: Went back and added the skill override, then refactored it substantially so the adjusted skill is returned in a separate function and carried around for the failure case.

**Comment by mlangsdorf** [comment]
catabindings is autogenerated. so why isn't it being autgenerated correctly?

File: src/bionics.cpp [hunk 10]
```diff
@@ -1535,14 +1497,18 @@ void bionic::deserialize( JsonIn &jsin )
...
+void player::introduce_into_anesthesia( time_duration const &duration,
+                                        bool external_kit )
 {
     add_msg_if_player( m_info,
                        _( "You type data into the console, configuring Autodoc to work with a CBM." ) );
     add_effect( effect_narcosis, duration );
     fall_asleep( duration );
     add_msg_if_player( m_info,
                        _( "Autodoc injected you with anesthesia, and while you were sleeping conducted a medical operation on you." ) );
+    if( external_kit ) {
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: What is this?

File: src/npctalk.cpp [hunk 11]
```diff
@@ -654,6 +659,10 @@ std::string dialogue::dynamic_line( const talk_topic &the_topic ) const
                    _( "&You are deaf and can't talk. When you don't respond, %s becomes angry!" ),
                    beta->name.c_str() );
     }
+    if( topic == "TALK_SEDATED" ) {
+        return string_format( _( "%s is sedated and can't wake up until it wears off" ),
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]: What "it"? Sedation or narcosis or anesthesia.

File: src/bionics.cpp [hunk 12]
```diff
@@ -996,35 +993,37 @@ bool player::install_bionics( const itype &type, int skill_level, bool autodoc )
...
+        if( is_player() ) {
+            add_memorial_log( pgettext( "memorial_male", "Installed bionic: %s." ),
+                              pgettext( "memorial_female", "Installed bionic: %s." ),
+                              bioid->name.c_str() );
+        }
         if( bioid->upgraded_bionic ) {
             remove_bionic( bioid->upgraded_bionic );
-            //~ %1$s - name of the bionic to be upgraded (inferior), %2$s - name of the upgraded bionic (superior).
+            //~ %1$s - name of the bionic to be upgraded (inferior),
+            //  %2$s - name of the upgraded bionic (superior).
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: Are you sure translation comments shouldn't be on one line? Here and below.

File: src/bionics.cpp (hunk 10 above)
Comment (mlangsdorf) [reply src/bionics.cpp]: if you've already collected the anesthesia kit in another routine, there's no need to search for it here.
The logic of the change is right and necessary, the implementation is not great. I'll change that to pass the item_comp or item_loc or something like that.

File: src/npctalk.cpp (hunk 11 above)
Comment (mlangsdorf) [reply src/npctalk.cpp]: Fixed.

File: src/bionics.cpp (hunk 12 above)
Comment (mlangsdorf) [reply src/bionics.cpp]: Fixed.

File: src/bionics.cpp (hunk 10 above)
Comment (mlangsdorf) [reply src/bionics.cpp]: going back to the boolean, because although it's a little inelegant, it at least works. I cannot figure out how to pass an item_location to another function because I am not very good at c++ but I am very frustrated.

File: src/bionics.cpp [hunk 13]
```diff
@@ -776,14 +774,14 @@ void player::process_bionic( int b )
     }
 }
 
-void bionics_uninstall_failure( player *u )
+void player::bionics_uninstall_failure( std::string installer_name )
 {
     switch( rng( 1, 5 ) ) {
         case 1:
-            add_msg( m_neutral, _( "You flub the removal." ) );
+            add_msg( m_neutral, _( "%s flub the removal." ), installer_name );
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: In case `%s` is a NPC, it will be "flubs", "messes" and "screws". Maybe change it to `add_msg_if_player_or_npc` here and below?

File: src/bionics.cpp [hunk 14]
```diff
@@ -857,12 +868,13 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
...
 
     if( !has_bionic( b_id ) ) {
-        popup( _( "You don't have this bionic installed." ) );
+        popup( _( "%s don't have this bionic installed." ), disp_name() );
         return false;
     }
 
     if( b_id == "bio_blaster" ) {
-        popup( _( "Removing your Fusion Blaster Arm would leave you with a useless stump." ) );
+        popup( _( "Removing %s Fusion Blaster Arm would leave %s with a useless stump." ),
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: I suppose first `%s` could be omitted and replaced with simple "the".

File: src/bionics.cpp [hunk 15]
```diff
@@ -875,81 +887,75 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
...
-                   b_id->name.c_str() );
+            popup( _( "%s must remove the %s bionic to remove the %s." ), installer.disp_name(),
+                   e.second.name.c_str(), b_id->name.c_str() );
             return false;
         }
     }
 
     if( b_id == "bio_eye_optic" ) {
-        popup( _( "The Telescopic Lenses are part of your eyes now.  Removing them would leave you blind." ) );
+        popup( _( "The Telescopic Lenses are part of %s eyes now.  Removing them would leave %s blind." ),
```
Comment (Night-Pryanik) [inline src/bionics.cpp]: I suppose first `%s` could be omitted and replaced with simple "the".

File: src/iexamine.cpp [hunk 16]
```diff
@@ -3474,6 +3477,51 @@ void iexamine::climb_down( player &p, const tripoint &examp )
     g->m.creature_on_trap( p );
 }
 
+player &player_on_couch( player &p, const tripoint autodoc_loc, player &null_patient, bool &adjacent_couch )
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Did you test moving out or destroying of autodoc couch? What's the point of having `adjacent_count` as a parameter? The code below doesn't check if it's true or false.

File: src/iexamine.cpp [hunk 17]
```diff
@@ -3483,29 +3531,30 @@ void iexamine::autodoc( player &p, const tripoint &examp )
...
         popup( _( "No patient found located on the connected couches.  Operation impossible.  Exiting." ) );
         return;
     }
 
-    const bool has_anesthesia = p.crafting_inventory().has_item_with( []( const item &it ) {
-        return it.has_flag( "ANESTHESIA" );
-    } );
+    const item_location anesthesia_kit = g->inv_map_splice( []( const item &e ) {
+                return e.has_flag( "ANESTHESIA" );
+    }, _( "Choose anesthesia kit" ), PICKUP_RANGE,
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Proposing to choose an anesthesia kit is an added keypress. Is there a real need for this not to be automatic?

File: src/iexamine.cpp [hunk 18]
```diff
@@ -3530,45 +3584,44 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             const itype *itemtype = it->type;
             const bionic_id &bid = itemtype->bionic->id;
 
-            if( p.has_bionic( bid ) ) {
-                popup( _( "You have already installed this bionic."  ) );
+            //~ %1$s is installer name and shouldn't need translation, %2$s is bionic CBM display name, %3$s is patient name and shouldn't need translation
+            add_msg( _( "%1$s prepares to install the %2$s on %3$s." ), installer.name,
+                     it->display_name().c_str(), patient.name );
+            if( patient.has_bionic( bid ) ) {
+                //~ %1$s is the You have/<npcname> has construct described above
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Translators on Transifex won't understand what sort of "construct described above" is meant here.

File: src/iexamine.cpp [hunk 19]
```diff
@@ -3516,6 +3565,11 @@ void iexamine::autodoc( player &p, const tripoint &examp )
 
     amenu.query();
 
+    // Pop-up messages ted to be of the form "You have $SOMETHING" or "NPC has $SOMETHING"
+    //~ Put the You have / NPC has section into a substring that only needs to be translated once
+    std::string pronoun = ( patient.name == p.name ? "You have " :
+                                                     string_format( _( "%1$s has "), patient.name ) );
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: I don't find this sort of construct appropriate. Other languages may have another word ordering. It would be better to replace `popup` with `add_msg_if_player_or_npc` and add two variants of messages, for the sake of convenience of translation.

File: src/npctalk.cpp [hunk 20]
```diff
@@ -1567,7 +1576,6 @@ void dialogue::gen_responses( const talk_topic &the_topic )
         }
 
     } else if( topic == "TALK_MISSION_DESCRIBE" ) {
-        add_response( _( "What's the matter?" ), "TALK_MISSION_OFFER" );
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]: Why did you remove this topic?

File: src/player.h [hunk 21]
```diff
@@ -270,15 +270,26 @@ class player : public Character
...
         /** Handles process of introducing patient into anesthesia through the autodoc */
-        void introduce_into_anesthesia( time_duration const &duration );
+        void introduce_into_anesthesia( time_duration const &duration, player &installer );
         /** Returns true if the player is wearing an active optical cloak */
         bool has_active_optcloak() const;
         /** Adds a bionic to my_bionics[] */
         void add_bionic(bionic_id const &b);
         /** Removes a bionic from my_bionics[] */
         void remove_bionic(bionic_id const &b);
+	/** Calculate skill for (un)installing bionics */
```
Comment (Night-Pryanik) [inline src/player.h]: Indent.

File: src/player.h [hunk 22]
```diff
@@ -270,15 +270,26 @@ class player : public Character
...
+                                      bool autodoc, int skill_level = -1 );
+        /** Attempts to install bionics, returns false if the player cancels prior to installation */
+        bool install_bionics( const itype &type, player &installer, bool autodoc = false,
+                              int skill_level = -1 );
+        void bionics_install_failure( player &installer, int difficulty, int success, float adjusted_skill );
         /** Used by the player to perform surgery to remove bionics and possibly retrieve parts */
-        bool uninstall_bionic(bionic_id const &b_id, int skill_level = -1, bool autodoc = false );
+        bool uninstall_bionic( bionic_id const &b_id, player &installer, bool autodoc = false,
+                               int skill_level = -1 );
+	void bionics_uninstall_failure( std::string installer_name );
```
Comment (Night-Pryanik) [inline src/player.h]: Indent.

**Comment by Night-Pryanik** [comment]
Why is there still only one commit after all your changes?

File: src/iexamine.cpp (hunk 16 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: adjacent_couch is assigned true during the search, and is checked on line 3539.

File: src/iexamine.cpp (hunk 18 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: I am learning I do not understand the translation support system. will fix.

File: src/iexamine.cpp (hunk 17 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: will fix as soon as I figure out how.

File: src/npctalk.cpp (hunk 20 above)
Comment (mlangsdorf) [reply src/npctalk.cpp]: messed that up, will fix.

**Comment by mlangsdorf** [comment]
I've been doing --amends because I don't think this code is ready for prime time.

File: src/iexamine.cpp (hunk 17 above)
Comment (Night-Pryanik) [reply src/iexamine.cpp]: It is automatic before. First the game checked for kit available, and then the game consumed it automatically through the `consume_items` function.

File: src/iexamine.cpp (hunk 16 above)
Comment (Night-Pryanik) [reply src/iexamine.cpp]: I talked about `adjacent_couch` as a parameter of `player_on_couch` function. If `adjacent_couch` is a parameter, it should be checked inside the function.

Comment (mlangsdorf) [reply src/iexamine.cpp]: It is passed as false into the function, the function sets it true as a side effect of finding the couch. The point of having it as a paramter is to alter the value if the couch is adjacent to the autodoc.

File: src/iexamine.cpp (hunk 17 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Right, but player:: introduce_to_anesthesia is now called by the patient, who may be an NPC without anesthesia in his inventory. So I had to move the code to push the comp up to here.

File: src/iexamine.cpp (hunk 19 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Removed.

File: src/iexamine.cpp (hunk 17 above)
Comment (Night-Pryanik) [reply src/iexamine.cpp]: Then why don't check for anesthesia kit in NPC inventory?

File: src/iexamine.cpp (hunk 16 above)
Comment (Night-Pryanik) [reply src/iexamine.cpp]: I think I got it. Though I think all this strange (for me) mechanics would not be needed if you haven't extracted `player_on_couch` function. I'm not a good programmer, but I intuitively feel that this is wrong, e.g.when function will do all it should do only when one of its parameters has a strictly specific name.

Comment (mlangsdorf) [reply src/iexamine.cpp]: I'm not 100% wild about it either, but:
patient is declared as a reference (player &patient), so it needs to be defined when it is declared.
patient is used throughout iexamine::autodoc, so I need to declare patient at the top scope of the function, and I can't declare it inside the if ( g->m.furn( couch_loc  ) == couch ) { scope.
I would rather not find the autodoc couch twice, so player_on_couch() should find the couch and the PC or NPC that is on it.

I could declare the return of player_on_couch() as std::pair<player &, bool> but that's honestly more hassle than it's worth comparing to just changing a parameter value as a side effect.

File: src/iexamine.cpp (hunk 17 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Because I think the use flow of "push the NPC onto the couch, activate autodoc, use CBM and kit from your inventory" is a lot more intuitive than "push the NPC onto the couch, give the NPC a CBM and anesthesia kit, activate autodoc, use CBM and kit from the NPC's inventory."

File: src/iexamine.cpp [hunk 23]
```diff
@@ -3535,45 +3596,57 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             const itype *itemtype = it->type;
             const bionic_id &bid = itemtype->bionic->id;
 
-            if( p.has_bionic( bid ) ) {
-                popup( _( "You have already installed this bionic."  ) );
+            if( patient.is_npc() && !bid->npc_install ) {
+                //~ %2$s is the bionic CBM display name, %2$s is the patient name 
+                popup( _( "%1$s cannot be installed on npc %2$s"), it->display_name().c_str(), patient.name );
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: It's better to remove "npc" part from the text.

File: src/player_display.cpp [hunk 24]
```diff
@@ -363,6 +384,16 @@ Strength - 4;    Dexterity - 4;    Intelligence - 4;    Perception - 4" ) );
             mvwputch( w_grid_trait, i, 26, BORDER_COLOR, LINE_XOXO );
         }
 
+        //Horizontal line bionics
+        if( i <= 26 ) {
+            mvwputch( w_grid_bionics, bionics_win_size_y, i, BORDER_COLOR, LINE_OXOX );
```
Comment (Night-Pryanik) [inline src/player_display.cpp]: Just in case: we have `mvwhline` for horizontal lines and `mvwvline` for vertical ones.

**Comment by Night-Pryanik** [comment]
Always forgetting about NPC surgeons. What skill sets they use to (un)install bionics after all your changes?

File: src/iexamine.cpp (hunk 23 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Okay.

File: src/player_display.cpp (hunk 24 above)
Comment (mlangsdorf) [reply src/player_display.cpp]: I'm just copying the code in player_display at this point; it clearly needs to be refactored a bunch but beyond the scope of this PR.

**Comment by mlangsdorf** [comment]
I put the skill override code back in, though I didn't special case it in the install_bionics code. If you pass a skill_level that is not -1 into bionics_adjusted_skill(), pl_skill is set to 12 * skill_level.

Simpler, no fuss, and no recalculation of adjusted skill in the failure path.

**Comment by mlangsdorf** [comment]
Closing this and replacing it with individual PRs for specific issues.
